### DISCUSSION TITLE
unable to invoke layer creator, Unable to find built artifact in target/*.[ejw]ar

### DISCUSSION THREAD

**Discussion by Rvarman0110** [paketo-buildpacks/feedback#227] [Java Team]
unable to invoke layer creator
unable to contribute application layer
unable to resolve artifacts
unable to find any built artifacts for pattern(s):/target/*.[ejw]ar

and in my pom.xml there is a modules section where we refer other pom.xml files which are located in the subfolders of main directory
<modules>
    <module>libraries</module>
    <module>starters</module>
</modules>
.
├── pom.xml
├── api
│ ├── pom.xml
│ └── target
└── web
├── pom.xml
├── target

now while building the build ends up successful but it is unable to find any jar there

(say there are 2 modules in my pom.xml then it will build those 2 jars and 1 main jar total 3 jars and it is unable to find those jars from there)

I tried using BP_MAVEN_BUILT_ARTIFACT and set the location of jar explicitly /source/code/target/*.[ejw]ar
but it still showing the same error how to resolve this issue (its a multi-module java project)

say in one of my projects I have a single module mentioned in pom.xml file like libraries and it has its own pom.xml, source code everything 
now while running I could able to see that the  modules jar gets build in a location 

[INFO] Building archetype jar: /source/code/libraries/target/libraries-0.2.0.jar

So I mentioned this path in my BP MAVEN BUILT ARTIFACT=/source/code/libraries/target/*.jar

now I am getting an error like 
unable to invoke layer creator
unable to contribute application layer
unable to resolve artifacts
unable to find any built artifacts for pattern(s):
/source/code/libraries/target/*.jar

**Comment by Rvarman0110** [comment]
Got it the problem is resolved as I was trying to mention the absolute path it couldnt able to find anything now I tried with the Relative path 
like BP MAVEN BUILT ARTIFACT=libraries/target/*.jar
now its working

**Comment by Rvarman0110** [comment]
But while setting this BP MAVEN BUILT ARTIFACT I was able to move to the exporting stage when i set process type as null but  getting an error when i use process type as web 
ERROR: failed to export: determining entrypoint: tried to set web to default but it doesn't exist

**Reply by dmikusa** [reply]
It sounds like nothing is setting a process type. This is done by different buildpacks depending on what type of app you have.

- For a WAR, it would be one of our app server buildpacks like Apache Tomcat, Apache Tomee, or Libery.
- For an executable JAR, it would be the executable-jar buildpack.
- We also support some other app types too, like dist-zip.

One of these buildpacks should run, so look at the output of your build and check to see if the one you'd expect is running. The output of the buildpack will indicate when it adds process types. You'll see them listed.

The other possibility is that the `BP MAVEN BUILT ARTIFACT` you're setting isn't specific enough or matching the correct file. You can have issues if it matches the wrong JAR/WAR or if it matches multiple JAR/WAR files.

If possible, post your full build output and build command. It's hard to say what's happening without that.

**Reply by Rvarman0110** [reply]
Hi 
 
the command that I used pack build application --env BP_JVM_VERSION=17
--env "BP_MAVEN_BUILD_ARTIFACT=libraries/target/libraries*.jar"  --builder paketobuildpacks/builder:base  --path . 

===> DETECTING
10 of 26 buildpacks participating
paketo-buildpacks/ca-certificates   3.6.1
paketo-buildpacks/bellsoft-liberica 10.2.3
paketo-buildpacks/syft              1.30.0
paketo-buildpacks/maven             6.15.2
paketo-buildpacks/executable-jar    6.7.1
paketo-buildpacks/apache-tomcat     7.13.3
paketo-buildpacks/apache-tomee      1.7.1
paketo-buildpacks/liberty           3.7.1
paketo-buildpacks/dist-zip          5.6.1
paketo-buildpacks/spring-boot       5.25.0
===> RESTORING
===> BUILDING
Paketo Buildpack for CA Certificates 3.6.1
  https://github.com/paketo-buildpacks/ca-certificates
  Launch Helper: Contributing to layer
    Creating /layers/paketo-buildpacks_ca-certificates/helper/exec.d/ca-certificates-helper
Paketo Buildpack for BellSoft Liberica 10.2.3
  https://github.com/paketo-buildpacks/bellsoft-liberica
  Build Configuration:
    $BP_JVM_JLINK_ARGS           --no-man-pages --no-header-files --strip-debug --compress=1  configure custom link arguments (--output must be omitted)
    $BP_JVM_JLINK_ENABLED        false                                                        enables running jlink tool to generate custom JRE
    $BP_JVM_TYPE                 JRE                                                          the JVM type - JDK or JRE
    $BP_JVM_VERSION              17                                                           the Java version
  Launch Configuration:
    $BPL_DEBUG_ENABLED           false                                                        enables Java remote debugging support
    $BPL_DEBUG_PORT              8000                                                         configure the remote debugging port
    $BPL_DEBUG_SUSPEND           false                                                        configure whether to suspend execution until a debugger has attached
    $BPL_HEAP_DUMP_PATH                                                                       write heap dumps on error to this path
    $BPL_JAVA_NMT_ENABLED        true                                                         enables Java Native Memory Tracking (NMT)
    $BPL_JAVA_NMT_LEVEL          summary                                                      configure level of NMT, summary or detail
    $BPL_JFR_ARGS                                                                             configure custom Java Flight Recording (JFR) arguments
    $BPL_JFR_ENABLED             false                                                        enables Java Flight Recording (JFR)
    $BPL_JMX_ENABLED             false                                                        enables Java Management Extensions (JMX)
    $BPL_JMX_PORT                5000                                                         configure the JMX port
    $BPL_JVM_HEAD_ROOM           0                                                            the headroom in memory calculation
    $BPL_JVM_LOADED_CLASS_COUNT  35% of classes                                               the number of loaded classes in memory calculation
    $BPL_JVM_THREAD_COUNT        250                                                          the number of threads in memory calculation
    $JAVA_TOOL_OPTIONS                                                                        the JVM launch flags
    Using Java version 17 from BP_JVM_VERSION
  BellSoft Liberica JDK 17.0.7: Contributing to layer
    Downloading from https://github.com/bell-sw/Liberica/releases/download/17.0.7+7/bellsoft-jdk17.0.7+7-linux-amd64.tar.gz
    Verifying checksum
    Expanding to /layers/paketo-buildpacks_bellsoft-liberica/jdk
    Adding 124 container CA certificates to JVM truststore
    Writing env.build/JAVA_HOME.override
    Writing env.build/JDK_HOME.override
  BellSoft Liberica JRE 17.0.7: Contributing to layer
    Downloading from https://github.com/bell-sw/Liberica/releases/download/17.0.7+7/bellsoft-jre17.0.7+7-linux-amd64.tar.gz
    Verifying checksum
    Expanding to /layers/paketo-buildpacks_bellsoft-liberica/jre
    Adding 124 container CA certificates to JVM truststore
    Writing env.launch/BPI_APPLICATION_PATH.default
    Writing env.launch/BPI_JVM_CACERTS.default
    Writing env.launch/BPI_JVM_CLASS_COUNT.default
    Writing env.launch/BPI_JVM_SECURITY_PROVIDERS.default
    Writing env.launch/JAVA_HOME.default
    Writing env.launch/JAVA_TOOL_OPTIONS.append
    Writing env.launch/JAVA_TOOL_OPTIONS.delim
    Writing env.launch/MALLOC_ARENA_MAX.default
  Launch Helper: Contributing to layer
    Creating /layers/paketo-buildpacks_bellsoft-liberica/helper/exec.d/active-processor-count
    Creating /layers/paketo-buildpacks_bellsoft-liberica/helper/exec.d/java-opts
    Creating /layers/paketo-buildpacks_bellsoft-liberica/helper/exec.d/jvm-heap
    Creating /layers/paketo-buildpacks_bellsoft-liberica/helper/exec.d/link-local-dns
    Creating /layers/paketo-buildpacks_bellsoft-liberica/helper/exec.d/memory-calculator
    Creating /layers/paketo-buildpacks_bellsoft-liberica/helper/exec.d/security-providers-configurer
    Creating /layers/paketo-buildpacks_bellsoft-liberica/helper/exec.d/jmx
    Creating /layers/paketo-buildpacks_bellsoft-liberica/helper/exec.d/jfr
    Creating /layers/paketo-buildpacks_bellsoft-liberica/helper/exec.d/security-providers-classpath-9
    Creating /layers/paketo-buildpacks_bellsoft-liberica/helper/exec.d/debug-9
    Creating /layers/paketo-buildpacks_bellsoft-liberica/helper/exec.d/nmt
    Creating /layers/paketo-buildpacks_bellsoft-liberica/helper/exec.d/openssl-certificate-loader
  Java Security Properties: Contributing to layer
    Writing env.launch/JAVA_SECURITY_PROPERTIES.default
    Writing env.launch/JAVA_TOOL_OPTIONS.append
    Writing env.launch/JAVA_TOOL_OPTIONS.delim
Paketo Buildpack for Syft 1.30.0
  https://github.com/paketo-buildpacks/syft
    Downloading from https://github.com/anchore/syft/releases/download/v0.80.0/syft_0.80.0_linux_amd64.tar.gz
    Verifying checksum
    Writing env.build/SYFT_CHECK_FOR_APP_UPDATE.default
Paketo Buildpack for Maven 6.15.2
  https://github.com/paketo-buildpacks/maven
  Build Configuration:
    $BP_EXCLUDE_FILES                                                                                    colon separated list of glob patterns, matched source files are removed
    $BP_INCLUDE_FILES                                                                                    colon separated list of glob patterns, matched source files are included
    $BP_JAVA_INSTALL_NODE                 false                                                          whether to install Yarn/Node binaries based on the presence of a package.json or yarn.lock file
    $BP_MAVEN_ACTIVE_PROFILES                                                                            the active profiles (comma separated: such as: p1,!p2,?p3) to pass to Maven
    $BP_MAVEN_ADDITIONAL_BUILD_ARGUMENTS                                                                 the additionnal arguments (appended to BP_MAVEN_BUILD_ARGUMENTS) to pass to Maven
    $BP_MAVEN_BUILD_ARGUMENTS             -Dmaven.test.skip=true --no-transfer-progress package          the arguments to pass to Maven
    $BP_MAVEN_BUILT_ARTIFACT              libraries/target/libraries*.jar                                the built application artifact explicitly.  Supersedes $BP_MAVEN_BUILT_MODULE
    $BP_MAVEN_BUILT_MODULE                                                                               the module to find application artifact in
    $BP_MAVEN_DAEMON_ENABLED              false                                                          use maven daemon
    $BP_MAVEN_POM_FILE                    pom.xml                                                        the location of the main pom.xml file, relative to the application root
    $BP_MAVEN_SETTINGS_PATH               settings.xml                                                   the path to a Maven settings file
    $BP_MAVEN_VERSION                     3                                                              the Maven version
    $BP_NODE_PROJECT_PATH                                                                                configure a project subdirectory to look for `package.json` and `yarn.lock` files
  Apache Maven 3.9.2: Contributing to layer
    Downloading from https://repo1.maven.org/maven2/org/apache/maven/apache-maven/3.9.2/apache-maven-3.9.2-bin.tar.gz
    Verifying checksum
    Expanding to /layers/paketo-buildpacks_maven/maven
    Creating cache directory /home/cnb/.m2
  Compiled Application: Contributing to layer
    Executing mvn --settings=settings.xml --batch-mode -Dmaven.test.skip=true --no-transfer-progress package
      [INFO] Scanning for projects...
      [INFO] ------------------------------------------------------------------------
      [INFO] Reactor Build Order:
      [INFO] 
      [INFO] scripts                                               [pom]
      [INFO] libraries                              [maven-archetype]
      [INFO] 
      [INFO] ----------------< scripts >----------------
      [INFO] Building scripts 0.3.0-local-SNAPSHOT                 [1/2]
      [INFO]   from pom.xml
      [INFO] --------------------------------[ pom ]---------------------------------
      [INFO] 
      [INFO] --- jacoco:0.8.7:prepare-agent (pre-unit-test) @ scripts ---
      [INFO] argLine set to -javaagent:/home/cnb/.m2/repository/org/jacoco/org.jacoco.agent/0.8.7/org.jacoco.agent-0.8.7-runtime.jar=destfile=/workspace/source/code/target/jacoco.exec
      [INFO] 
      [INFO] --- jacoco:0.8.7:report (post-unit-test) @ scripts ---
      [INFO] Skipping JaCoCo execution due to missing execution data file.
      [INFO] 
      [INFO] --- source:3.2.1:jar-no-fork (attach-sources) @ scripts ---
      [INFO] 
      [INFO] --- class-graph:1.1.15-693d3ce:class-loader-overlap (default) @ scripts ---
      [info] Project packaging is not jar but pom. Nothing to do.
      [INFO] 
      [INFO] --------------< libraries>--------------
      [INFO] Building libraries 0.3.0-local-SNAPSHOT             [2/2]
      [INFO]   from libraries/pom.xml
      [INFO] --------------------------[ maven-archetype ]---------------------------
      [INFO] 
      [INFO] --- jacoco:0.8.7:prepare-agent (pre-unit-test) @ libraries---
      [INFO] argLine set to -javaagent:/home/cnb/.m2/repository/org/jacoco/org.jacoco.agent/0.8.7/org.jacoco.agent-0.8.7-runtime.jar=destfile=/workspace/source/code/libraries/target/jacoco.exec
      [INFO] 
      [INFO] --- resources:3.3.1:resources (default-resources) @ libraries---
      [INFO] Copying 2 resources from src/main/resources to target/classes
      [INFO] Copying 36 resources from src/main/resources to target/classes
      [INFO] 
      [INFO] --- resources:3.3.1:testResources (default-testResources) @ libraries---
      [INFO] Not copying test resources
      [INFO] 
      [INFO] --- jacoco:0.8.7:report (post-unit-test) @ libraries---
      [INFO] Skipping JaCoCo execution due to missing execution data file.
      [INFO] 
      [INFO] --- archetype:3.2.0:jar (default-jar) @ libraries---
      [INFO] Building archetype jar: /workspace/source/code/libraries/target/libraries-0.3.0-local-SNAPSHOT.jar
      [INFO] 
      [INFO] Building jar: /workspace/source/code/libraries/target/libraries-0.3.0-local-SNAPSHOT.jar
      [INFO] 
      [INFO] --- source:3.2.1:jar-no-fork (attach-sources) @ libraries---
      [INFO] Building jar: /workspace/source/code/libraries/target/libraries-0.3.0-local-SNAPSHOT-sources.jar
      [INFO] 
      [INFO] --- ds-devops-class-graph:1.1.15-20220112T191512-693d3ce:class-loader-overlap (default) @ libraries---
      [info] Project packaging is not jar but maven-archetype. Nothing to do.
      [INFO] ------------------------------------------------------------------------
      [INFO] Reactor Summary for scripts 0.3.0-local-SNAPSHOT:
      [INFO] 
      [INFO] scripts ............................... SUCCESS [  8.127 s]
      [INFO] libraries........................... SUCCESS [  4.229 s]
      [INFO] ------------------------------------------------------------------------
      [INFO] BUILD SUCCESS
      [INFO] ------------------------------------------------------------------------
      [INFO] Total time:  02:19 min
      [INFO] Finished at: 2023-11-23T15:49:04Z
      [INFO] ------------------------------------------------------------------------
      [WARNING] 
      [WARNING] Plugin validation issues were detected in 4 plugin(s)
      [WARNING] 
      [WARNING]  * org.apache.maven.plugins:maven-source-plugin:3.2.1
      [WARNING]  * org.jacoco:jacoco-maven-plugin:0.8.7
      [WARNING]  * devops-class-graph:1.1.15-693d3ce
      [WARNING]  * org.apache.maven.plugins:maven-archetype-plugin:3.2.0
      [WARNING] 
      [WARNING] For more or less details, use 'maven.plugin.validation' property with one of the values (case insensitive): [BRIEF, DEFAULT, VERBOSE]
      [WARNING] 
      
  Removing source code
  Restoring multiple artifacts
===> EXPORTING
Adding layer 'paketo-buildpacks/ca-certificates:helper'
Adding layer 'paketo-buildpacks/bellsoft-liberica:helper'
Adding layer 'paketo-buildpacks/bellsoft-liberica:java-security-properties'
Adding layer 'paketo-buildpacks/bellsoft-liberica:jre'
Adding layer 'buildpacksio/lifecycle:launch.sbom'
Adding 1/1 app layer(s)
Adding layer 'buildpacksio/lifecycle:launcher'
Adding layer 'buildpacksio/lifecycle:config'
Adding label 'io.buildpacks.lifecycle.metadata'
Adding label 'io.buildpacks.build.metadata'
Adding label 'io.buildpacks.project.metadata'
ERROR: failed to export: determining entrypoint: tried to set web to default but it doesn't exist

Step failed

**Reply by dmikusa** [reply]
Yes, the problem is that your `BP_MAVEN_BUILD_ARTIFACT ` isn't specific enough. It matches multiple artifacts, and then no other buildpacks run because nothing knows how to execute what was pulled out of the build. Because no other buildpacks run, nothing ends up setting a process type. When the export happens, it expects a process type of `web` to exist, it doesn't so that fails.

You can see it's matching multiple with the line `Restoring multiple artifacts`. Then it never runs any more buildpacks, but it should run something. How would you expect your app to run? Is it an executable JAR? Do you need an app server? Also, what is your Maven build producing? a WAR, a JAR, a set of JAR files?
